src/ln/channelmanager.rs

```diff
@@ -2077,152 +2121,92 @@ impl ChannelManager {
-														let reported_cltv_expiry = byte_utils::slice_to_be32(&err_packet.failuremsg[2..2+4]);
-														onion_failure_log!("incorrect_cltv_expiry", UPDATE|13, "cltv_expiry", reported_cltv_expiry);
-														route_hop.cltv_expiry_delta as u16 >= chan_update.contents.cltv_expiry_delta
-													},
-													c if c == UPDATE|20 => { // channel_disabled
-														let reported_flags = byte_utils::slice_to_be16(&err_packet.failuremsg[2..2+2]);
-														onion_failure_log!("channel_disabled", UPDATE|20, "flags", reported_flags);
-														chan_update.contents.flags & 0x01 == 0x01
-													},
-													c if c == UPDATE|21 => true, // expiry_too_far
-													_ => { unreachable!(); },
-												};
-
-												let msg = if is_chan_update_invalid { None } else {
-													Some(msgs::HTLCFailChannelUpdate::ChannelUpdateMessage {
-														msg: chan_update,
-													})
-												};
-												res = Some((msg, true));
-												return;
-											}
+							let mut fail_channel_update = None;
+
+							if error_code & NODE == NODE {
+								fail_channel_update = Some(msgs::HTLCFailChannelUpdate::NodeFailure { node_id: route_hop.pubkey, is_permanent: error_code & PERM == PERM });
+							}
+							else if error_code & PERM == PERM {
+								fail_channel_update = if payment_failed {None} else {Some(msgs::HTLCFailChannelUpdate::ChannelClosed {
+								  short_channel_id: match error_code & 0xff {
+								    8|9|10 => route.hops[next_route_hop_ix].short_channel_id, // outgoing (permanent_channel_failure/required_channel_feature_missing/unknown_next_peer)
+								    _ => route_hop.short_channel_id, // incoming channel (invalid_realm & BADONION)
+								  },
+								   is_permanent: error_code & PERM == PERM,
+								})};
+							}
+							else if error_code & UPDATE == UPDATE {
+								if err_packet.failuremsg.len() >= debug_field_size + 2 {
+									let update_len = byte_utils::slice_to_be16(&err_packet.failuremsg[debug_field_size+2..debug_field_size+4]) as usize;
+									if err_packet.failuremsg.len() >= debug_field_size + 4 + update_len {
+										if let Ok(chan_update) = msgs::ChannelUpdate::read(&mut Cursor::new(&err_packet.failuremsg[debug_field_size + 4..debug_field_size + 4 + update_len])) {
+											// if channel_update should NOT have caused the failure:
+											// MAY treat the channel_update as invalid.
+											let is_chan_update_invalid = match error_code & 0xff {
+												7 => false,
+												11 => amt_to_forward > chan_update.contents.htlc_minimum_msat,
+												12 => {
+													let new_fee = amt_to_forward.checked_mul(chan_update.contents.fee_proportional_millionths as u64).and_then(|prop_fee| { (prop_fee / 1000000).checked_add(chan_update.contents.fee_base_msat as u64) });
+													new_fee.is_none() || incoming_htlc_msat >= new_fee.unwrap() && incoming_htlc_msat >= amt_to_forward + new_fee.unwrap()
```

Comment:
I think new_fee.is_none() is a valid update-causes failure, no? It indicates the node is setting their feerate to some mega-high value causing an overflow, which probably means they're temporarily disabling the channel by setting a high fee (or at least disabling high-value forwards).

Reply:
Also, why the second condition (incoming_htlc_msat >= new_fee.unwrap()), why not just amt_to_forward.checked_add(new_fee.unwrap()) and compare?